htmresearch/algorithms/sparsematrix_temporal_memory/basal_context_apical_context.py

```diff
@@ -0,0 +1,723 @@
+    (learningMatchingBasalSegments,
+     learningMatchingApicalSegments) = self._chooseBestSegmentPairPerColumn(
+       matchingCellsInBurstingColumns, matchingBasalSegments,
+       matchingApicalSegments, basalPotentialExcitations,
+       apicalPotentialExcitations)
+    newSegmentCells = self._getCellsWithFewestSegments(
+      burstingColumnsWithNoMatch)
+
+    # Incorrectly predicted columns
+    if self.predictedSegmentDecrement > 0.0:
+      correctMatchingBasalMask = np.in1d(
+        cellsForMatchingBasal / self.cellsPerColumn, activeColumns)
+      correctMatchingApicalMask = np.in1d(
+        cellsForMatchingApical / self.cellsPerColumn, activeColumns)
+
+      punishedBasalSegments = matchingBasalSegments[~correctMatchingBasalMask]
+      punishedApicalSegments = matchingApicalSegments[~correctMatchingApicalMask]
+    else:
+      punishedBasalSegments = ()
+      punishedApicalSegments = ()
+
+    # Make a list of every cell that is learning
+    learningCells =  np.concatenate(
+      (correctPredictedCells,
+       self.basalConnections.mapSegmentsToCells(learningMatchingBasalSegments),
+       newSegmentCells))
+
+    return (learningActiveBasalSegments,
+            learningActiveApicalSegments,
+            learningMatchingBasalSegments,
+            learningMatchingApicalSegments,
+            punishedBasalSegments,
+            punishedApicalSegments,
+            newSegmentCells,
+            learningCells)
+
+
+  def _calculateSegmentActivity(self, lateralInput, feedbackInput):
+    """
+    Calculate the active and matching segments for this timestep.
+
+    @param lateralInput (numpy array)
+    @param feedbackInput (numpy array)
+
+    @return (tuple)
+    - activeBasalSegments (numpy array)
+      Basal dendrite segments with enough active connected synapses to cause a
+      dendritic spike
+
+    - activeApicalSegments (numpy array)
+      Apical dendrite segments with enough active connected synapses to cause a
+      dendritic spike
+
+    - matchingBasalSegments (numpy array)
+      Basal dendrite segments with enough active potential synapses to be
+      selected for learning in a bursting column
+
+    - matchingApicalSegments (numpy array)
+      Apical dendrite segments with enough active potential synapses to be
+      selected for learning in a bursting column
+
+    - basalPotentialExcitations (numpy array)
+      The number of active potential synapses for each basal segment.
+      Includes counts for active, matching, and nonmatching segments.
+
+    - apicalPotentialExcitations (numpy array)
+      The number of active potential synapses for each apical segment
+      Includes counts for active, matching, and nonmatching segments.
+    """
+
+    basalPermanences = self.basalConnections.matrix
+    apicalPermanences = self.apicalConnections.matrix
+
+    # Active basal
+    basalExcitations = basalPermanences.rightVecSumAtNZGteThresholdSparse(
```

Comment:
Consider naming as "basalSegmentOverlaps"
Consider helper function in segmentSparseMatrix for this